### ISSUE TITLE
Integration into and existing app

### ISSUE THREAD

**Issue by DavidKmn** [mentrena/SyncKit#143]
Hello Manuel @mentrena, great work together with the rest on the library. Our team was planning on using SyncKit in our project that uses CoreData. Our use case is to create backups for user data (every 24 hours) and allow users to retrieve these backups on new devices they have. 

A couple of integration issues we have:
1) We already setup our own stack with our own NSPersistentContainer and NSPersistentStoreCoordinator. Looking at the example we see that there is a new NSPersistentStoreCoordinator being setup for SyncKit, so is it fine to have 2 NSPersistentStoreCoordinators? (I know that it is possible however not sure if thats the right approach)
2) Not all core data models we have require syncing, about 3 out of the 15 models we have. Is is possible to limit the syncing scope to these models only?
3) We see the usage of deviceId in the library however what if we have multiple user accounts on the same device? How would a use case like this be supported?
4) We have 3 entities which have interlinked relationships, the idea was to sync them atomically as it does not make sense to upload them separately. I see that the ModifyRecordsOperation does not allow for this in SyncKit, what are your thoughts on this?

**Comment by mentrena** [comment]
Hi David,

1. It is totally fine to have more than one Core Data stack in your app, and the way SyncKit keeps track of changes is by using its own Core Data model. In your app you would set up your NSPersistentContainer normally and then pass the managed object context to SyncKit, and your app wouldn't need to worry about how SyncKit keeps its own internal data.
2. At the moment that's not possible for the Core Data implementation, but you could use two separate Core Data stacks in your app, one that gets synchronized with iCloud and one that's entirely local.
3. That would indeed be complicated, keep in mind iCloud uses the device Apple ID to identify users, so even if your app allows separate accounts the device might still be using the same iCloud account. My suggestion would be to wipe out user data on logout, or, if that's not an option, to use some sort of `userId` in the path for your NSPersistentStore, so you will keep separate stores for different user accounts –but in that case you will also need to wind up the core data stack and SyncKit synchronizer on logout, and bring them up again on login. This problem lies within the scope of your application though, it's not specific to SyncKit. 
That deviceId you mention would not affect your use case, it's just an optimisation to avoid downloading data that was uploaded from the same device.
4. That is a limitation that CloudKit has. The documentation for `CKModifyRecordsOperation` says:
_To ensure the speed of fetching and saving records, the server may reject large operations. When this occurs, a block reports the CKError.Code.limitExceeded error. Your app should handle this error, and refactor the operation into multiple smaller batches._
If you want to make sure they are uploaded together you could do something like assign a relatively small batchSize to the synchronizer to make "sure" CloudKit will never complain about it (something like 50), and then make sure to call `synchronize()` when you know that there are fewer changes than that number. Although maybe that doesn't fit your use case of synchronising only once a day. And again there's nothing *sure*, you might try to upload just a few records and CloudKit might still send the `limitExceeded` error, the record limit in a batch is not really specified anywhere.

Hope that helps!

**Comment by DavidKmn** [comment]
Thanks for the response!
1. Are we talking about a background context we would create with NSPersistentContaner.newBackgroundContext()? Don't see where this would be passed into SyncKit..
2. This may be a bit difficult  since we have cross references with other entities and we do custom mapping before the models are sent off. Is there no way to use a custom implementation of the ModelProvider (from my understanding the layers can be seperated)
3. Got it!
4. Got it!

**Comment by DavidKmn** [comment]
Looking more into the code I found this like SyncKit does a separate fetch for each object that is going to be uploaded during syncronization. Seems like this is a bit inefficient as there will be quite a lot of IO operations.. Why was a batched fetch not implemented to fetch 200-300 objects at a time and use fetchLimits fetchOffsets to go through the database? 

We have 100,000 of objects and this would seem like a bottleneck at this point unless I am missing an implementation detail? Would love to hear your though and work on a solution if an issue is present.

![Screenshot 2021-02-20 at 15 33 38](https://user-images.githubusercontent.com/35953014/108600799-01a56200-7391-11eb-9197-9d396ce2ce67.png)

**Comment by mentrena** [comment]
> Are we talking about a background context we would create with NSPersistentContaner.newBackgroundContext() and pass in?

You would pass the context that you want SyncKit to observe and merge changes to.

> Go this to work by doing custom work with core data configurations. Can share lated in a PR if its somethings others would like to do.

You mean on the side of your Core Data stack or by making changes to the SyncKit code?

> We were actually thinking of doing a separate zone for each user in cloudkit, so there would be essentially a zone perfix with the user id, would this have downsides we are missing?

I guess that would work, but the question is more specific to the domain of your application. My first instinct as a user, if I'm logged in and using an app, is that the app will show my data as long as I've logged in successfully. In this case you would _also_ require that the two devices are using the same AppleID so they have access to the same record zone. If that fits your use case then it's all good.

> Looking more into the code I found this like SyncKit does a separate fetch for each object that is going to be uploaded during syncronization.

I created a ticket to do this one year ago, some day I'll get to do it 😅 
It's just not a priority for me, my apps where I use SyncKit typically sync after any changes happen, so there's rarely a case of loads of records needing to be uploaded at the same time.
